### DISCUSSION TITLE
Dynamic service providers and arguments

### DISCUSSION THREAD

**Discussion by maryam4s26** [os-js/OS.js#846] [Q&A] [answered]
hi @andersevenrud 

In your opinion, how can I control the registration of adapters and service providers through the `docker-compose.yml` file?

**Comment by andersevenrud** [comment, marked as answer]
I would suggest using (prefixed) environmental flags if you just want basic bootstrapping.

For example imagine you have two providers: Foo and Bar. 

```js
const enabledServices = process.env.OSJS_SERVICES.split(',')

if (enabledServices.includes('Foo')) {
  osjs.register(FooServiceProvider, {
    args: {
      arg1: process.env.OSJS_SERVICE_FOO_ARG1,
      arg2: process.env.OSJS_SERVICE_FOO_ARG2,
    }
  })
}
if (enabledServices.includes('Bar')) {
  osjs.register(BarServiceProvicer, {
    args: {
      arg1: process.env.OSJS_SERVICE_BAR_ARG1,
      arg2: process.env.OSJS_SERVICE_BAR_ARG2,
    }
  })
}
```

Then this could be set up with:

```
OSJS_SERVICES=Foo,Bar
OSJS_SERVICE_FOO_ARG1=Hello
OSJS_SERVICE_FOO_ARG2=World
OSJS_SERVICE_BAR_ARG1=Bye
OSJS_SERVICE_BAR_ARG2=World
```

---

If you need something more advanced, use something like a JSON file that you bind via a [ro] volume and just read that in your bootstrap file and do whatever you need. This can then be copied into an image in your Dockerfile or whatever for deployments.

**Comment by andersevenrud** [comment]
Just a note on dynamically registering service providers and behaviour on runtime.

To make sure things don't crash whenever a service provider has not been registered and you perform `core.make('my-namespace/something')` calls, etc.:

* Define the provided calls in your `provides()`
* Do `if (core.has()) {}` wrapping where you do the actual calls

---

```js
class MyServiceProvider {
  constructor(core, options = {}) {
    this.core = core;
    this.options = options;
  }

  provides() {
    return ['my-namespace/something'];
  }

  async init() {
    this.core.singleton('my-namespace/something', () => ({
      foo: name => alert(`Hello ${name}!`)
    }));
  }
}
```

```js
if (core.has('my-namespace/something')) {
  const {foo} = core.make('my-namespace/something')
  foo('World')
}
```

**Comment by andersevenrud** [comment]
And I just had a thought. You can actually set up the client-side providers via the backend so you don't have to compile to apply settings from the environment.

1. Create a provider on the server that sets up an endpoint that returns some kind of configuration object based on `process.env` like described in my first post
2. On the client bootstrap `init` function, do a HTTP call to the above endpoint
3. Now you can register anything like you want it

---

```js
// Your server provider
class MyServiceProvider {
  constructor(core, options = {}) {
    this.core = core;
    this.options = options;
  }

  async init() {
    const {route} = this.core.make('osjs/express');

    route('GET', '/my-endpoint', (req, res) => {
      res.json({ services: ['Foo', 'Bar'] });
    });
  }
}
```

```js
// Your client bootstrap
async function init() {
  const myConfiguration = await osjs.request('/my-endpoint');
  // ... register stuff
  // --> myConfiguration.services
}
```

**Comment by maryam4s26** [comment]
Thank you for your complete explanation <3

**Comment by andersevenrud** [comment]
You're very welcome!
